testing/citest/tests/kube_v2_smoke_test.py

```diff
@@ -0,0 +1,267 @@
+                'kind': 'Deployment',
+                'metadata': {
+                    'name': name,
+                    'namespace': self.TEST_NAMESPACE,
+                    'labels': {
+                        'app': self.TEST_APP,
+                    }
+                },
+                'spec': {
+                    'replicas': 1,
+                    'selector': {
+                        'matchLabels': {
+                            'app': self.TEST_APP,
+                        }
+                    },
+                    'template': {
+                        'metadata': {
+                            'labels': {
+                                'app': self.TEST_APP,
+                            }
+                        },
+                        'spec': {
+                            'containers': [{
+                                'name': 'primary',
+                                'image': self.__desired_image_id,
+                            }]
+                        }
+                    }
+                }
+            }]
+        }],
+        description='Deploy manifest',
+        application=self.TEST_APP)
+
+    builder = kube.KubeContractBuilder(self.kube_v2_observer)
+    (builder.new_clause_builder('Deployment created',
+                                retryable_for_secs=15)
+     .get_resources(
+         'deploy',
+         extra_args=[name, '--namespace', self.TEST_NAMESPACE])
+     .contains_path_eq('spec/replicas', 1))
+
+    return st.OperationContract(
+        self.new_post_operation(
+            title='deploy_manifest', data=payload, path='tasks'),
+        contract=builder.build())
+
+  def delete_manifest(self):
+    """Creates OperationContract for deleteManifest
+
+    To verify the operation, we just check that the Kubernetes deployment
+    is no longer visible (or is in the process of terminating).
+    """
+    bindings = self.bindings
+    name = self.TEST_APP + '-deployment'
+    payload = self.agent.make_json_payload_from_kwargs(
+        job=[{
+            'cloudProvider': 'kubernetes',
+            'type': 'deleteManifest',
+            'account': bindings['SPINNAKER_KUBERNETES_V2_ACCOUNT'],
+            'user': '[anonymous]',
+            'kinds': [ 'deployment' ],
+            'location': self.TEST_NAMESPACE,
+            'options': { },
+            'labelSelectors': {
+                'selectors': [{
+                    'kind': 'EQUALS',
+                    'key': 'app',
+                    'values': [ self.TEST_APP ]
+                }]
+            }
+        }],
+        application=self.TEST_APP,
+        description='Destroy Manifest')
+
+    builder = kube.KubeContractBuilder(self.kube_v2_observer)
+    (builder.new_clause_builder('Replica Set Removed')
+     .get_resources(
+         'deployment',
+         extra_args=[name, '--namespace', self.TEST_NAMESPACE],
+         no_resource_ok=True)
+     .contains_path_eq('targetSize', 0))
```

Comment:
Should this accomodate 404 errors as well?
See the error handling in the kube smoke test.

Reply:
I just noticed the no_resource_ok=True -- that is deprecated and I'm hoping to remove it in favor of the more generalized and standardizable response handling via EXPECT and ov_factory.error_list predicates.

Reply:
Can you share an example usage?

Reply:
from the kube_smoke_test.py

      (builder.new_clause_builder('Daemonset Deleted', retryable_for_secs=15)
       .get_resources(
          'daemonsets', extra_args=[self.TEST_APP + self.POSTFIX_STATEFUL_DAEMONSET,
                        '--namespace', self.TEST_NAMESPACE])
       .EXPECT(ov_factory.value_list_path_contains('targetSize', jp.NUM_EQ(0)))
       .OR(self.__not_found_observation_predicate()))

where
ov_factory = jc.ObservationPredicateFactory()

and

  def __not_found_observation_predicate(self, title='Not Found Permitted'):
    return ov_factory.error_list_contains(
        st.CliAgentRunErrorPredicate(
            title=title, error_regex='.* not found'))

What this is doing is using explicit predicates on the observation (the ov_factory helper) rather than predicates on the observation value. The ov_factory helper provides convienence to say whether you mean on the observation "value list" to look for good results, or on the "error list" to look for bad results. Being explicit this way, lets you say you are expecting to observe a 404 error (though the agent doesnt give 404's so we regex the output text) or a result with a 0 size. The "no_resources_ok" is simpler syntax but more wishy washy, harder to implement internally, and doesnt generalize to other types of errors.

I'd like a simpler syntax but havent realized one yet. So for the meantime I've been using that pattern so if I do find a simpler syntax, it will be straightforward to refactor to it and for the time being, the consistency and familiarity should make it a little easier to read.

Reply:
Oh nice this seems like exactly what I needed -- thanks!